### DISCUSSION TITLE
Is there a way to know what the original working directory of the terminal that launched cake?

### DISCUSSION THREAD

**Discussion by mattleibow** [cake-build/cake#3076] [Q&A] [answered]
I store all my cake files out of the root to keep things organised. But not only that, I sometimes use the cake files from different locations.

Some of my parameters are relative paths, and they will never resolve.

Is ther a way to know where cake was originally launched from so I can combine paths?

**Comment by augustoproiete** [comment]
Hi @mattleibow,

When you run a Cake script, the working directory is always set to the folder where the entry-point Cake script is located. So the relative path to other Cake scripts would have to be relative to that particular folder.

If I understand your question, you have other scripts in other folders, that are `#load`ed by other scripts also in different folders, so the relative path cannot always be known - Is that correct?

Assuming that's the case, this is not something Cake supports out-of-the-box. Even if you were able to find out the original working directory, the `#load` directive is preprocessed before you have any chance to execute any code.

One ~workaround~ hack :see_no_evil:  you could try, is to use a [bootstrapper script](https://github.com/cake-build/resources/tree/master/dotnet-tool) (e.g. `.ps1`, `.cmd`, `.sh`) to launch Cake, which would give you the opportunity to capture the working directory of the place where the bootstrapper was launched from, and then pass it to the Cake runner as an argument.

![image](https://user-images.githubusercontent.com/177608/106998531-09db8b80-675b-11eb-8ef7-5680747caab5.png)

This would give you the ability to receive the original working directory in your Cake script e.g.

```csharp
var launchedFrom = Argument<string>("launchedFrom", null);

if (!string.IsNullOrWhiteSpace(launchedFrom))
{
    Information("This Cake script was launched from {0}", launchedFrom);
}
```

You'd have to adapt each of the [bootstrappers](https://github.com/cake-build/resources/tree/master/dotnet-tool) that you use to capture the working directory and pass it as an argument.

For PowerShell, you could use `Get-Location` or `$PSScriptRoot` depending on what you want to do. The former gives you the folder where you launched the `build.ps1` from and the latter gives you the folder where the `build.ps1` is located.

Now the good news is that you have a way of getting the original working directory from within your Cake script. The bad news is that it's too late for the `#load` directive (which was already processed before your code executed), however you can still combine paths to other Cake scripts and run them via [`CakeExecuteScript`](https://cakebuild.net/api/Cake.Common.Tools.Cake/CakeAliases/) which will execute the Cake script at run time.

```csharp
CakeExecuteScript(script, new CakeSettings { Arguments = arguments });
```

This means that you might have to change some of your scripts to be "runnable" rather than "loadable" if that makes sense.

Of course, you can also get creative and copy files around, make changes, and execute on-the-fly via `CakeExecuteScript`. You can see a similar [example here](https://github.com/cake-contrib/Cake.Recipe/blob/2.1.0/Cake.Recipe/Content/tools.cake#L42-L47).

**Reply by mattleibow** [reply]
Thanks for this. Since the script locations are all known and do not change, the loads will be fine. I was more thinking about if there was some property on the context on the launch location.

However, if this is not a feature, then I can use the bootstrapper idea. Maybe we should add a property to CakeContext?

**Reply by augustoproiete** [reply]
> However, if this is not a feature, then I can use the bootstrapper idea. Maybe we should add a property to CakeContext?

I think it's a good idea. `Context.Environment` currently exposes `WorkingDirectory` which matches the directory where the `.cake` script is executing from. We could add another property e.g. `LaunchDirectory` or `LaunchedFromDirectory`

@cake-build/cake-team what do you think?

**Reply by patriksvensson** [reply]
I think it sounds reasonable to add a `LaunchDirectory` property to the Environment. Just as we make it obvious that this is not the working directory.

**Reply by devlead** [reply]
`LaunchDirectory` sounds like reasonable addition to me too.

**Comment by augustoproiete** [comment, marked as answer]
@mattleibow tracking via https://github.com/cake-build/cake/issues/3092
PRs welcome!
